// @hvisions/app-core/lib/components/Profile/index.js
"use strict";

require("core-js/modules/es.symbol");

require("core-js/modules/es.array.concat");

require("core-js/modules/es.array.index-of");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.object.get-own-property-descriptor");

require("core-js/modules/es.object.keys");

require("core-js/modules/es.object.to-string");

require("core-js/modules/es.string.iterator");

require("core-js/modules/es.weak-map");

require("core-js/modules/web.dom-collections.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _fUi = require("@hvisions/f-ui");

var _session = require("@hvisions/toolkit/lib/session");

var _style = _interopRequireDefault(require("./style.scss"));

var _Core = _interopRequireDefault(require("../../Core"));

var types = _interopRequireWildcard(require("../../store/core/types"));

var _reactRedux = require("react-redux");

var _session2 = require("../../utils/session");

var _avatar = _interopRequireDefault(require("./imgs/avatar.png"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Profile = function Profile(_ref) {
  var f7router = _ref.f7router,
      props = _objectWithoutProperties(_ref, ["f7router"]);

  var _dispatch = (0, _reactRedux.useDispatch)();

  var logout = function logout() {
    _dispatch({
      type: types.CHANGE_MODULE_LIST,
      moduleList: []
    });

    _dispatch({
      type: types.CHANGE_HOME_MENU,
      homeMenus: []
    });

    (0, _session2.clear)();
    f7router.navigate('/login');
  };

  var showAbout = function showAbout() {
    f7router.app.dialog.create({
      title: '关于',
      content: "<div class=".concat(_style.default.aboutContent, ">\n        <p>").concat(_Core.default.getHomeConfig().version || '版本号v2.0', "</p>\n        <p>@ Copyright 2022</p>\n        <p>Shanghai H VISIONS Technology</p>\n        <p>Co.,Ltd.</p>\n      </div>"),
      buttons: [{
        text: '确定',
        cssClass: _style.default.confirmBtn
      }]
    }).open();
  };

  return _react.default.createElement("div", {
    className: _style.default.profilePage
  }, _react.default.createElement("div", {
    className: _style.default.avatarImage
  }, _react.default.createElement("img", {
    src: _avatar.default,
    alt: ""
  })), _react.default.createElement("span", {
    className: _style.default.username
  }, (0, _session.getAuthData)() && (0, _session.getAuthData)().userName), _react.default.createElement(_fUi.Card, {
    className: _style.default.cardContainer
  }, _react.default.createElement(_fUi.List, null, _react.default.createElement(_fUi.ListItem, {
    header: "\u8D26\u53F7",
    title: (0, _session.getAuthData)() && (0, _session.getAuthData)().userAccount
  })), _react.default.createElement(_fUi.List, null, _react.default.createElement(_fUi.ListItem, {
    header: "\u90E8\u95E8",
    title: (0, _session.getAuthData)() && (0, _session.getAuthData)().departmentName
  })), _react.default.createElement(_fUi.List, null, _react.default.createElement(_fUi.ListItem, {
    header: "\u624B\u673A\u53F7",
    title: (0, _session.getAuthData)() && (0, _session.getAuthData)().mobilePhone
  })), _react.default.createElement(_fUi.List, null, _react.default.createElement(_fUi.ListItem, {
    header: "\u90AE\u7BB1",
    title: (0, _session.getAuthData)() && (0, _session.getAuthData)().email
  }))), _react.default.createElement(_fUi.Button, {
    className: _style.default.aboutBtn,
    onClick: showAbout
  }, "\u5173\u4E8E"), _react.default.createElement(_fUi.Button, {
    className: _style.default.logoutBtn,
    onClick: logout
  }, "\u9000\u51FA\u767B\u5F55"));
};

var _default = Profile;
exports.default = _default;
