// @hvisions/app-core/lib/components/Layout/Component/index.js
"use strict";

require("core-js/modules/es.symbol");

require("core-js/modules/es.array.concat");

require("core-js/modules/es.array.fill");

require("core-js/modules/es.array.filter");

require("core-js/modules/es.array.index-of");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.array.map");

require("core-js/modules/es.object.get-own-property-descriptor");

require("core-js/modules/es.object.keys");

require("core-js/modules/es.object.to-string");

require("core-js/modules/es.string.iterator");

require("core-js/modules/es.weak-map");

require("core-js/modules/web.dom-collections.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _style = _interopRequireDefault(require("./style.scss"));

var _fUi = require("@hvisions/f-ui");

var _session = require("@hvisions/toolkit/lib/session");

var _homeTitle = _interopRequireDefault(require("./imgs/homeTitle.png"));

var _role = _interopRequireDefault(require("./imgs/role.png"));

var _user = _interopRequireDefault(require("./imgs/user.png"));

var _reactRedux = require("react-redux");

var _lodash = require("lodash");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Layout = function Layout(_ref) {
  var moduleList = _ref.moduleList,
      props = _objectWithoutProperties(_ref, ["moduleList"]);

  var _core = (0, _reactRedux.useSelector)(function (_ref2) {
    var core = _ref2.core;
    return core;
  });

  var renderHomeCard = function renderHomeCard() {
    var rootPages = moduleList.filter(function (i) {
      return i.parentId == 0;
    });
    return _react.default.createElement(_fUi.Row, null, rootPages.concat(Array((3 - rootPages.length % 3) % 3).fill({})).map(function (item, index) {
      if ((0, _lodash.isEmpty)(item)) {
        return _react.default.createElement(_fUi.Col, {
          key: index,
          width: "33"
        });
      }

      return _react.default.createElement(_fUi.Col, {
        key: index,
        width: "33"
      }, _react.default.createElement("div", {
        className: _style.default.navigateCard,
        key: index,
        style: {
          textAlign: 'center'
        },
        onClick: function onClick() {
          props.$f7router.navigate(item.urlAddress);
        }
      }, _react.default.createElement("div", {
        className: _style.default.largeIcon
      }, _react.default.createElement("i", {
        className: item.icon ? "app-icon-".concat(item.icon, " ").concat(_style.default.iconContainer2) : _style.default.iconContainer2
      })), _react.default.createElement("div", {
        className: _style.default.cardFullName
      }, item.fullName)));
    }));
  };

  var renderAllFunctions = function renderAllFunctions() {
    var allFuncList = _core.homeMenus;
    return _react.default.createElement(_fUi.Row, null, allFuncList.concat({
      urlAddress: '/menu-more',
      fullName: '更多',
      iconIos: 'menu-more',
      // f7自带icon
      icon: 'menu-more'
    }).concat(Array((4 - (allFuncList.length + 1) % 4) % 4).fill({})).map(function (item, index) {
      if ((0, _lodash.isEmpty)(item)) {
        return _react.default.createElement(_fUi.Col, {
          width: "25",
          key: index
        });
      }

      return _react.default.createElement(_fUi.Col, {
        width: "25",
        key: index
      }, _react.default.createElement("div", {
        className: _style.default.navigateCard,
        key: index,
        style: {
          textAlign: 'center'
        },
        onClick: function onClick() {
          props.$f7router.navigate(item.urlAddress);
        }
      }, _react.default.createElement("i", {
        className: item.icon ? "app-icon-".concat(item.icon, " ").concat(_style.default.iconContainer) : _style.default.iconContainer
      }), _react.default.createElement("div", {
        className: _style.default.cardFullName
      }, item.fullName)));
    }));
  };

  return _react.default.createElement("div", {
    className: _style.default.homePage
  }, _react.default.createElement("div", {
    className: _style.default.homeTitle
  }, _react.default.createElement("img", {
    src: _homeTitle.default,
    slot: "media",
    alt: "",
    style: {
      width: 156
    }
  }), _react.default.createElement("div", {
    style: {
      textAlign: 'center'
    },
    className: _style.default.subTitle
  }, "\u5236\u9020\u534F\u540C\u5E73\u53F0", ' ')), _react.default.createElement("div", {
    className: _style.default.welcome
  }, _react.default.createElement("div", {
    className: _style.default.welcomeText
  }, "\u60A8\u597D\uFF0C", (0, _session.getAuthData)() && (0, _session.getAuthData)().userName, "!"), _react.default.createElement("div", {
    className: _style.default.userCard
  }, _react.default.createElement(_fUi.Card, {
    className: _style.default.leftCard
  }, _react.default.createElement(_fUi.CardContent, null, _react.default.createElement("div", {
    className: _style.default.cardTitle
  }, _react.default.createElement("div", {
    className: _style.default.imgContainer
  }, _react.default.createElement("img", {
    src: _user.default,
    alt: ""
  })), _react.default.createElement("span", null, "\u8D26\u53F7")), _react.default.createElement("div", {
    className: _style.default.cardContent
  }, (0, _session.getAuthData)() && (0, _session.getAuthData)().userAccount))), _react.default.createElement(_fUi.Card, {
    className: _style.default.rightCard
  }, _react.default.createElement(_fUi.CardContent, null, _react.default.createElement("div", {
    className: _style.default.cardTitle
  }, _react.default.createElement("div", {
    className: _style.default.imgContainer
  }, _react.default.createElement("img", {
    src: _role.default,
    alt: ""
  })), _react.default.createElement("span", null, "\u90E8\u95E8")), _react.default.createElement("div", {
    className: _style.default.cardContent
  }, (0, _session.getAuthData)() && (0, _session.getAuthData)().departmentName)))), 
  
  _react.default.createElement("div", {
    className: _style.default.navigate2
  }, _react.default.createElement(_fUi.BlockTitle, null, "\u7CFB\u7EDF\u5BFC\u822A"), (0, _lodash.isEmpty)(moduleList) && _react.default.createElement(_fUi.Block, {
    strong: true,
    inset: true,
    className: "skeleton-text skeleton-effect-blink"
  }, _react.default.createElement(_fUi.Row, null, Array(4).fill({}).map(function (module, index) {
    return _react.default.createElement(_fUi.Col, {
      width: "25",
      key: index
    }, _react.default.createElement("div", {
      className: _style.default.navigateCard
    }, _react.default.createElement(_fUi.SkeletonBlock, {
      style: {
        width: '30px',
        height: '30px'
      },
      slot: "media"
    }), _react.default.createElement("div", {
      className: "skeleton-text"
    }, "aaaaaaa")));
  }))), !(0, _lodash.isEmpty)(moduleList) && _react.default.createElement(_fUi.Block, {
    strong: true,
    inset: true
  }, renderHomeCard()))));
};

var _default = Layout;
exports.default = _default;
