// @hvisions/app-core/lib/components/Layout/Component/MenuMoreSkeleton.js
"use strict";

require("core-js/modules/es.array.fill");

require("core-js/modules/es.array.map");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _fUi = require("@hvisions/f-ui");

var _style = _interopRequireDefault(require("./style.scss"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var skeleton = function skeleton(props) {
  return _react.default.createElement("div", {
    className: _style.default.navigate
  }, _react.default.createElement(_fUi.BlockTitle, {
    className: "skeleton-text"
  }, "aaaa"), _react.default.createElement(_fUi.Block, {
    strong: true,
    inset: true,
    className: "skeleton-text skeleton-effect-blink"
  }, _react.default.createElement(_fUi.Row, null, Array(4).fill({}).map(function (module, index) {
    return _react.default.createElement(_fUi.Col, {
      width: "25",
      key: index
    }, _react.default.createElement("div", {
      className: _style.default.navigateCard
    }, _react.default.createElement(_fUi.SkeletonBlock, {
      style: {
        width: '30px',
        height: '30px'
      },
      slot: "media"
    }), _react.default.createElement("div", {
      className: "skeleton-text"
    }, "aaaaaaa")));
  }))));
};

var _default = skeleton;
exports.default = _default;
