// @hvisions/app-core/lib/components/Layout/Component/MenuMore.js
"use strict";

require("core-js/modules/es.symbol");

require("core-js/modules/es.symbol.description");

require("core-js/modules/es.symbol.iterator");

require("core-js/modules/es.array.concat");

require("core-js/modules/es.array.fill");

require("core-js/modules/es.array.filter");

require("core-js/modules/es.array.find");

require("core-js/modules/es.array.flat");

require("core-js/modules/es.array.from");

require("core-js/modules/es.array.index-of");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.array.map");

require("core-js/modules/es.array.unscopables.flat");

require("core-js/modules/es.object.get-own-property-descriptor");

require("core-js/modules/es.object.keys");

require("core-js/modules/es.object.to-string");

require("core-js/modules/es.regexp.to-string");

require("core-js/modules/es.string.iterator");

require("core-js/modules/es.weak-map");

require("core-js/modules/web.dom-collections.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactRedux = require("react-redux");

var types = _interopRequireWildcard(require("../../../store/core/types"));

var _fUi = require("@hvisions/f-ui");

var _style = _interopRequireDefault(require("./style.scss"));

var _toolkit = require("@hvisions/toolkit");

var _lodash = require("lodash");

var _menu = _interopRequireDefault(require("../../../api/menu"));

var _MenuMoreSkeleton = _interopRequireDefault(require("./MenuMoreSkeleton"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * ### 导航栏设置
 * 过滤出所有一级菜单，配置底部Toolbar导航栏
 * @returns
 */
var MenuMore = function MenuMore(_ref) {
  var _ref$moduleList = _ref.moduleList,
      moduleList = _ref$moduleList === void 0 ? [] : _ref$moduleList,
      props = _objectWithoutProperties(_ref, ["moduleList"]);

  var _dispatch = (0, _reactRedux.useDispatch)();

  var _core = (0, _reactRedux.useSelector)(function (_ref2) {
    var core = _ref2.core;
    return core;
  });

  var _moduleTree = _toolkit.tree.formatTree(moduleList);

  var renderHomeMenus = function renderHomeMenus() {
    var homeMenuModules = _core.homeMenus;

    if ((0, _lodash.isEmpty)(homeMenuModules)) {
      homeMenuModules = ['placeholder'];
    }

    return _react.default.createElement(_fUi.Row, null, homeMenuModules.concat(Array((4 - homeMenuModules.length % 4) % 4).fill({})).map(function (item, index) {
      if (item == 'placeholder') {
        return _react.default.createElement(_fUi.Col, {
          width: "25",
          key: index
        }, _react.default.createElement("div", {
          className: _style.default.navigateCard,
          key: index,
          style: {
            textAlign: 'center'
          }
        }, _react.default.createElement("i", {
          className: _style.default.iconContainer
        }, _react.default.createElement(_fUi.Icon, {
          icon: "plus",
          f7: "plus",
          style: {
            border: '1px dashed',
            fontSize: '40px'
          }
        })), _react.default.createElement("div", {
          className: _style.default.cardFullName
        }, " ")));
      }

      if ((0, _lodash.isEmpty)(item)) {
        return _react.default.createElement(_fUi.Col, {
          width: "25",
          key: index
        });
      }

      return _react.default.createElement(_fUi.Col, {
        key: index,
        width: "25"
      }, _react.default.createElement("div", {
        className: _style.default.navigateCard,
        key: index,
        style: {
          textAlign: 'center'
        },
        onClick: function onClick() {
          return handleRemoveMenu(item);
        }
      }, _react.default.createElement("i", {
        className: item.icon ? "app-icon-".concat(item.icon, " ").concat(_style.default.iconContainer) : _style.default.iconContainer
      }), _react.default.createElement("div", {
        className: _style.default.cardFullName
      }, item.fullName), _react.default.createElement(_fUi.Badge, {
        color: "red",
        style: {
          position: 'absolute',
          top: '-8px',
          right: 8,
          width: 16,
          '--f7-badge-size': '18px'
        }
      }, _react.default.createElement(_fUi.Icon, {
        f7: "minus",
        style: {
          fontSize: 12
        }
      }))));
    }));
  };

  var handleRemoveMenu = function handleRemoveMenu(item) {
    var menuList = _core.homeMenus.filter(function (i) {
      return i.id != item.id;
    });

    _menu.default.saveMenu(menuList.map(function (i) {
      return i.id;
    }));

    _dispatch({
      type: types.CHANGE_HOME_MENU,
      homeMenus: menuList
    });
  };

  var handleAddMenu = function handleAddMenu(item) {
    if (!(0, _lodash.isEmpty)(_core.homeMenus.find(function (tempMenu) {
      return tempMenu.urlAddress == item.urlAddress;
    }))) {
      return;
    }

    if (_core.homeMenus.length >= 7) {
      return;
    }

    var menuList = [].concat(_toConsumableArray(_core.homeMenus), [item]);

    _menu.default.saveMenu(menuList.map(function (i) {
      return i.id;
    }));

    _dispatch({
      type: types.CHANGE_HOME_MENU,
      homeMenus: menuList
    });
  };

  var flatTree = function flatTree(treeItem) {
    if ((0, _lodash.isEmpty)(treeItem) || (0, _lodash.isEmpty)(treeItem.children)) {
      return [treeItem];
    }

    if (treeItem.parentId == 0) {
      return [].concat(treeItem.children.map(function (i) {
        return flatTree(i);
      }).flat(Infinity));
    }

    return [treeItem].concat(treeItem.children.map(function (i) {
      return flatTree(i);
    }).flat(Infinity));
  };

  var renderSubMenus = function renderSubMenus(submenu) {
    var submenuModules = flatTree(submenu).filter(function (i) {
      return (0, _lodash.isEmpty)(_core.homeMenus.find(function (tempMenu) {
        return tempMenu.urlAddress == i.urlAddress;
      }));
    });

    if ((0, _lodash.isEmpty)(submenuModules)) {
      submenuModules = Array(4).fill({});
    }

    return _react.default.createElement(_fUi.Row, null, submenuModules.concat(Array((4 - submenuModules.length % 4) % 4).fill({})).map(function (module, index) {
      if ((0, _lodash.isEmpty)(module)) {
        return _react.default.createElement(_fUi.Col, {
          width: "25",
          key: index
        }, _react.default.createElement("div", {
          className: _style.default.navigateCard,
          key: index,
          style: {
            textAlign: 'center'
          }
        }, _react.default.createElement("i", {
          className: _style.default.iconContainer
        }), _react.default.createElement("div", {
          className: _style.default.cardFullName
        }, " ")));
      }

      return _react.default.createElement(_fUi.Col, {
        key: index,
        width: "25"
      }, _react.default.createElement("div", {
        className: _style.default.navigateCard,
        key: index,
        style: {
          textAlign: 'center'
        },
        onClick: function onClick() {
          return handleAddMenu(module);
        }
      }, _react.default.createElement("i", {
        className: module.icon ? "app-icon-".concat(module.icon, " ").concat(_style.default.iconContainer) : _style.default.iconContainer
      }), _react.default.createElement("div", {
        className: _style.default.cardFullName
      }, module.fullName), _core.homeMenus.length < 7 && _react.default.createElement(_fUi.Badge, {
        style: {
          position: 'absolute',
          top: '-8px',
          right: 8,
          width: 16,
          '--f7-badge-size': '18px',
          '--f7-badge-bg-color': "#255CB2"
        }
      }, _react.default.createElement(_fUi.Icon, {
        f7: "plus",
        style: {
          fontSize: 12
        }
      }))));
    }));
  };

  var renderAllMenus = function renderAllMenus() {
    return _moduleTree.map(function (item, index) {
      return _react.default.createElement("div", {
        key: index,
        className: _style.default.navigate3
      }, _react.default.createElement(_fUi.BlockTitle, null, item.fullName), _react.default.createElement(_fUi.Block, {
        strong: true,
        inset: true
      }, renderSubMenus(item)));
    });
  };

  return _react.default.createElement("div", null, _react.default.createElement("div", {
    className: _style.default.navigate3
  }, _react.default.createElement(_fUi.Block, {
    strong: true,
    inset: true,
    style: {
      marginTop: '16px !important'
    }
  }, renderHomeMenus())), (0, _lodash.isEmpty)(moduleList) && Array(3).fill({}).map(function (i, index) {
    return _react.default.createElement(_MenuMoreSkeleton.default, {
      key: index
    });
  }), !(0, _lodash.isEmpty)(moduleList) && renderAllMenus());
};

var _default = MenuMore;
exports.default = _default;
