// @hvisions/app-core/lib/utils/session.js
"use strict";

require("core-js/modules/es.symbol");

require("core-js/modules/es.array.concat");

require("core-js/modules/es.array.filter");

require("core-js/modules/es.object.get-own-property-descriptor");

require("core-js/modules/es.object.get-own-property-descriptors");

require("core-js/modules/es.object.keys");

require("core-js/modules/web.dom-collections.for-each");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getToken = getToken;
exports.saveToken = saveToken;
exports.saveApiIP = saveApiIP;
exports.saveApiPort = saveApiPort;
exports.saveApiHttps = saveApiHttps;
exports.getAuthMenus = getAuthMenus;
exports.getAuthData = getAuthData;
exports.getApiIP = getApiIP;
exports.getApiPort = getApiPort;
exports.getApiHttps = getApiHttps;
exports.getApiAddress = getApiAddress;
exports.setAuthData = setAuthData;
exports.setAuthMenus = setAuthMenus;
exports.updateAuthData = updateAuthData;
exports.setLocale = setLocale;
exports.getLocale = getLocale;
exports.getIsMenuPermission = getIsMenuPermission;
exports.clear = clear;
exports.isLoggedIn = isLoggedIn;
exports.default = void 0;

var _lodash = require("lodash");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var AUTH_TOKEN_KEY = 'AUTH_TOKEN';
var AUTH_DATA_KEY = 'AUTH_DATA';
var AUTH_MENU = 'AUTH_MENU';
var API_IP_KEY = 'API_IP_KEY';
var API_PORT_KEY = 'API_PORT_KEY';
var API_HTTPS_KEY = 'API_HTTPS_KEY';

function getToken() {
  return localStorage.getItem(AUTH_TOKEN_KEY);
}

function saveToken(token) {
  return localStorage.setItem(AUTH_TOKEN_KEY, token);
}

function saveApiIP(params) {
  return localStorage.setItem(API_IP_KEY, params);
}

function saveApiPort(params) {
  return localStorage.setItem(API_PORT_KEY, params);
}

function saveApiHttps(params) {
  return localStorage.setItem(API_HTTPS_KEY, params);
}

function getAuthMenus() {
  var data;

  try {
    data = JSON.parse(localStorage.getItem(AUTH_MENU));
  } catch (e) {
    data = null;
  }

  return data;
}

function getAuthData() {
  var data;

  try {
    data = JSON.parse(localStorage.getItem(AUTH_DATA_KEY));
  } catch (e) {
    // do nothing
    data = null;
  }

  return data;
}

function getApiIP() {
  return localStorage.getItem(API_IP_KEY) || process.env.__API_HOST__ || '';
}

function getApiPort() {
  return localStorage.getItem(API_PORT_KEY) || process.env.__API_PORT__ || '';
}

function getApiHttps() {
  var https = localStorage.getItem(API_HTTPS_KEY) || process.env.__API_HTTPS__;

  return https === 'true';
}

function getApiAddress() {
  var isHttps = getApiHttps();
  var prefix = isHttps ? 'https' : 'http';
  var port = ":".concat(getApiPort());

  if (!isHttps && port == 80 || isHttps && port == 443) {
    // 默认端口就不需要加在后面了
    port = '';
  }

  return "".concat(prefix, "://").concat(getApiIP()).concat(port);
}

function setAuthData(data) {
  return localStorage.setItem(AUTH_DATA_KEY, JSON.stringify(data));
}

function setAuthMenus(data) {
  return localStorage.setItem(AUTH_MENU, JSON.stringify(data));
}

function updateAuthData(key, value) {
  if (!isLoggedIn()) {
    return false;
  }

  var data = getAuthData();

  if (data) {
    data = _objectSpread({}, data, _defineProperty({}, key, value));
    setAuthData(data);
  }
}

function setLocale(locale) {
  return localStorage.setItem('locale', locale);
}

function getLocale() {
  return localStorage.getItem('locale') || 'zh_CN';
}

function getIsMenuPermission(key) {
  var data = getAuthMenus();

  if (!(0, _lodash.isEmpty)(data)) {
    return (0, _lodash.isEmpty)(data.filter(function (item) {
      return item.fullName === key;
    }));
  }
}

function clear() {
  localStorage.removeItem(AUTH_DATA_KEY);
  localStorage.removeItem(AUTH_TOKEN_KEY);
}

function isLoggedIn() {
  var token = getToken();
  var data = getAuthData();
  return Boolean(token) && Boolean(data);
}

var _default = {
  getToken: getToken,
  saveToken: saveToken,
  getAuthData: getAuthData,
  setAuthData: setAuthData,
  updateAuthData: updateAuthData,
  clear: clear,
  isLoggedIn: isLoggedIn,
  getIsMenuPermission: getIsMenuPermission,
  getAuthMenus: getAuthMenus,
  setAuthMenus: setAuthMenus,
  getApiIP: getApiIP,
  getApiPort: getApiPort,
  getApiAddress: getApiAddress,
  getApiHttps: getApiHttps,
  saveApiIP: saveApiIP,
  saveApiPort: saveApiPort,
  saveApiHttps: saveApiHttps
};
exports.default = _default;
