// @hvisions/app-core/lib/components/Login/Component/index.js
"use strict";

require("core-js/modules/es.symbol");

require("core-js/modules/es.symbol.description");

require("core-js/modules/es.symbol.iterator");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.object.get-prototype-of");

require("core-js/modules/es.object.set-prototype-of");

require("core-js/modules/es.object.to-string");

require("core-js/modules/es.promise");

require("core-js/modules/es.string.iterator");

require("core-js/modules/web.dom-collections.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("regenerator-runtime/runtime");

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactRedux = require("react-redux");

var _fUi = require("@hvisions/f-ui");

var _toolkit = require("@hvisions/toolkit");

var _actions = require("../../../store/session/actions");

var _login = _interopRequireDefault(require("./login.scss"));

var _sys_logo = _interopRequireDefault(require("./sys_logo.png"));

var _ip = _interopRequireDefault(require("./ip.png"));

var _port = _interopRequireDefault(require("./port.png"));

var _Core = _interopRequireDefault(require("../../../Core"));

var _homeTitle = _interopRequireDefault(require("../../Layout/Component/imgs/homeTitle.png"));

var _lodash = require("lodash");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var getApiIP = _toolkit.session.getApiIP,
    getApiPort = _toolkit.session.getApiPort,
    getApiHttps = _toolkit.session.getApiHttps,
    saveApiIP = _toolkit.session.saveApiIP,
    saveApiPort = _toolkit.session.saveApiPort,
    saveApiHttps = _toolkit.session.saveApiHttps;

var Login =
/*#__PURE__*/
function (_React$Component) {
  _inherits(Login, _React$Component);

  function Login(props) {
    var _this;

    _classCallCheck(this, Login);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Login).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "signIn",
    /*#__PURE__*/
    _asyncToGenerator(
    /*#__PURE__*/
    regeneratorRuntime.mark(function _callee() {
      var self, app, router;
      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              self = _assertThisInitialized(_this);
              app = self.$f7;
              router = self.$f7router;
              _context.prev = 3;
              _context.next = 6;
              return _this.props.login({
                account: self.state.username,
                password: self.state.password,
                siteNum: self.state.siteNum
              });

            case 6:
              router.navigate('/');
              _context.next = 12;
              break;

            case 9:
              _context.prev = 9;
              _context.t0 = _context["catch"](3);
              app.dialog.alert(_context.t0.message);

            case 12:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, null, [[3, 9]]);
    })));

    _defineProperty(_assertThisInitialized(_this), "saveSetting", function () {
      var ipReg = /^(\d{1,2}|1\d\d|2[0-4]\d|25[0-5])\.(\d{1,2}|1\d\d|2[0-4]\d|25[0-5])\.(\d{1,2}|1\d\d|2[0-4]\d|25[0-5])\.(\d{1,2}|1\d\d|2[0-4]\d|25[0-5])$/;
      var hostReg = /^(?=^.{3,255}$)[a-zA-Z0-9][-a-zA-Z0-9]{0,62}(\.[a-zA-Z0-9][-a-zA-Z0-9]{0,62})+$/;

      if ((0, _lodash.isEmpty)(_this.state.ip) || (0, _lodash.isEmpty)(_this.state.port)) {
        _this.$f7.toast.create({
          position: 'center',
          text: 'ip和端口信息不完整',
          closeTimeout: '2000'
        }).open();

        return;
      }

      if (!ipReg.test(_this.state.ip) && !hostReg.test(_this.state.ip)) {
        _this.$f7.toast.create({
          position: 'center',
          text: 'ip信息不正确',
          closeTimeout: '2000'
        }).open();

        return;
      }

      saveApiIP(_this.state.ip);
      saveApiPort(_this.state.port); // saveApiHttps(this.state.useHttps);

      _this.setState({
        state: 'login'
      });
    });

    _this.state = {
      state: 'login',
      username: '',
      password: '',
      siteNum: '',
      ip: getApiIP(),
      port: getApiPort(),
      useHttps: getApiHttps()
    };
    return _this;
  }

  _createClass(Login, [{
    key: "renderForm",
    value: function renderForm() {
      var _this2 = this;

      if (this.state.state === 'login') {
        return _react.default.createElement(_fUi.List, {
          form: true
        }, _react.default.createElement(_fUi.ListInput, {
          label: "\u7528\u6237\u540D",
          type: "text",
          placeholder: "\u8BF7\u8F93\u5165\u7528\u6237\u540D",
          value: this.state.username,
          onInput: function onInput(e) {
            _this2.setState({
              username: e.target.value
            });
          },
          clearButton: true
        }), _react.default.createElement(_fUi.ListInput, {
          label: "\u5BC6\u7801",
          type: "password",
          placeholder: "\u8BF7\u8F93\u5165\u5BC6\u7801",
          value: this.state.password,
          onInput: function onInput(e) {
            _this2.setState({
              password: e.target.value
            });
          },
          clearButton: true
        }), _Core.default.getSaas() && _react.default.createElement(_fUi.ListInput, {
          type: "text",
          placeholder: "\u4F01\u4E1A\u7F16\u7801",
          value: this.state.siteNum,
          onInput: function onInput(e) {
            _this2.setState({
              siteNum: e.target.value
            });
          },
          clearButton: true
        }, _react.default.createElement(_fUi.Icon, {
          f7: "house",
          slot: "media",
          color: "white"
        })), _react.default.createElement(_fUi.ListInput, {
          className: _login.default.hide
        }));
      }

      return _react.default.createElement(_fUi.List, {
        form: true,
        className: _login.default.configForm
      }, _react.default.createElement(_fUi.ListInput, {
        type: "text",
        placeholder: "\u8BF7\u8F93\u5165ip(\u793A\u4F8B:127.0.0.1)",
        value: this.state.ip,
        onInput: function onInput(e) {
          _this2.setState({
            ip: e.target.value
          });
        },
        clearButton: true
      }, _react.default.createElement("img", {
        src: _ip.default,
        slot: "media",
        alt: "",
        style: {
          width: 28
        }
      })), _react.default.createElement(_fUi.ListInput, {
        type: "text",
        placeholder: "\u8BF7\u8F93\u5165\u7AEF\u53E3(\u793A\u4F8B:3000)",
        value: this.state.port,
        onInput: function onInput(e) {
          _this2.setState({
            port: e.target.value
          });
        },
        clearButton: true
      }, _react.default.createElement("img", {
        src: _port.default,
        slot: "media",
        alt: "",
        style: {
          width: 28
        }
      })));
    }
  }, {
    key: "renderBtn",
    value: function renderBtn() {
      var _this3 = this;

      if (this.state.state === 'login') {
        return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_fUi.Button, {
          large: true,
          raised: true,
          fill: true,
          round: true,
          onClick: this.signIn
        }, "\u767B\u5F55"), _react.default.createElement(_fUi.Button, {
          color: "gray",
          onClick: function onClick() {
            return _this3.setState({
              state: 'ip'
            });
          }
        }, "\u8BBE\u7F6E"));
      }

      return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement(_fUi.Button, {
        round: true,
        onClick: this.saveSetting
      }, "\u4FDD\u5B58"), _react.default.createElement(_fUi.Button, {
        color: "gray",
        onClick: function onClick() {
          return _this3.setState({
            ip: getApiIP(),
            port: getApiPort(),
            state: 'login'
          });
        }
      }, "\u8FD4\u56DE"));
    }
  }, {
    key: "render",
    value: function render() {
      return _react.default.createElement(_fUi.Page, {
        noToolbar: true,
        noNavbar: true,
        noSwipeback: true,
        loginScreen: true,
        className: _login.default.login
      }, _react.default.createElement("div", {
        className: _login.default.loginBg
      }, _react.default.createElement("header", {
        className: _login.default.logo
      }, _react.default.createElement("div", {
        className: _login.default.title
      }, _react.default.createElement("img", {
        src: _homeTitle.default,
        slot: "media",
        alt: "",
        style: {
          height: '100%'
        }
      })), _react.default.createElement("div", {
        className: _login.default.welcome
      }, "\u60A8\u597D\uFF0C\u6B22\u8FCE\u767B\u5F55\uFF01"), _react.default.createElement("div", {
        className: _login.default.hint
      }, "\u8BF7\u4F7F\u7528\u8D26\u53F7\u5BC6\u7801\u767B\u5F55")), _react.default.createElement("section", {
        className: _login.default.loginForm
      }, this.renderForm(), _react.default.createElement(_fUi.Block, {
        className: _login.default.loginBtn
      }, this.renderBtn())), _react.default.createElement("div", {
        className: _login.default.footer
      }, _react.default.createElement("div", null, _Core.default.getHomeConfig().version || '版本号v2.0'), _react.default.createElement("div", null, "\xA9", ' ', _Core.default.getHomeConfig().bottomDesc || 'Copyright 2022 Shanghai H VISIONS Technology Co., Ltd.'))));
    }
  }]);

  return Login;
}(_react.default.Component);

Login.propTypes = {
  login: _propTypes.default.func
};

var mapStateToProps = function mapStateToProps() {
  return {};
};

var _default = (0, _reactRedux.connect)(mapStateToProps, {
  login: _actions.login
})(Login);

exports.default = _default;
