// @hvisions/needle-service/template/src/store/task/types.js
export const GET_TASK = 'task/GET_TASK';
export const GET_DAYREATINTERVALTYPE = 'task/GET_DAYREATINTERVALTYPE';
export const GET_DAYREPEATINTERVALTYPEENUM = 'task/GET_DAYREPEATINTERVALTYPEENUM';
export const GET_DAYREPEATTYPEENUM = 'task/GET_DAYREPEATTYPEENUM';
export const GET_MONTHINTERVALTYPEENUM = 'task/GET_MONTHINTERVALTYPEENUM';
export const GET_REPEATTYPEENUM = 'task/GET_REPEATTYPEENUM';
export const GET_SINGLEDAYTYPEENUM = 'task/GET_SINGLEDAYTYPEENUM';
export const GET_SINGLEWEEKDAYENUM = 'task/GET_SINGLEWEEKDAYENUM';
export const GET_TASKCOUNT = 'task/GET_TASKCOUNT';
export const GET_PRIORITYNAME = 'task/GET_PRIORITYNAME';
export const GET_TASKBYTASKID = 'task/GET_TASKBYTASKID';
export const GET_SPAREBYPROCESSINSTANCEID = 'task/GET_SPAREBYPROCESSINSTANCEID';
export const GET_LUBBYPROCESSINSTANCEID = 'task/GET_LUBBYPROCESSINSTANCEID';
export const GET_LUBAPPLY = 'task/GET_LUBAPPLY';
export const GET_SPAREAPPLY = 'task/GET_SPAREAPPLY';
export const GETLUBAPPLYTHROUGH = 'task/GETLUBAPPLYTHROUGH';
export const GETSPAREAPPLYTHROUGH = 'task/GETSPAREAPPLYTHROUGH';
export const GET_HISTORIC = 'task/GET_HISTORIC';
export const GET_THISTORYPROCESSINSTANCEBYID = 'task/GET_THISTORYPROCESSINSTANCEBYID';

export const GETSPAREAPPLYLIST = 'task/GETSPAREAPPLYLIST';

export const GETLUBAPPLYLIST = 'task/GETLUBAPPLYLIST';
export const GET_HISTORICPROCESSINSTANCE = 'task/GET_HISTORICPROCESSINSTANCE';
