// @hvisions/app-core/lib/Core.js
"use strict";

require("core-js/modules/es.symbol");

require("core-js/modules/es.symbol.description");

require("core-js/modules/es.symbol.iterator");

require("core-js/modules/es.array.concat");

require("core-js/modules/es.array.filter");

require("core-js/modules/es.array.from");

require("core-js/modules/es.array.index-of");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.array.map");

require("core-js/modules/es.object.get-own-property-descriptor");

require("core-js/modules/es.object.get-own-property-descriptors");

require("core-js/modules/es.object.keys");

require("core-js/modules/es.object.to-string");

require("core-js/modules/es.regexp.to-string");

require("core-js/modules/es.string.iterator");

require("core-js/modules/web.dom-collections.for-each");

require("core-js/modules/web.dom-collections.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _lodash = require("lodash");

var _reactRedux = require("react-redux");

var _framework7LiteEsmBundle = _interopRequireDefault(require("framework7/framework7-lite.esm.bundle.js"));

var _fUi = _interopRequireDefault(require("@hvisions/f-ui"));

var _store = _interopRequireDefault(require("./store"));

var _plugin = require("./plugin");

var _toolkit = require("@hvisions/toolkit");

var _Root = _interopRequireDefault(require("./components/Root"));

var _Context = _interopRequireDefault(require("./components/Config/Context"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var modules = [];
var HomeConfigs = {};
var NextConfigs = {};
var isSaas = false; // func useModule
// export function use(module) {
//   modules.push(module);
// }

var Core =
/*#__PURE__*/
function () {
  function Core(options) {
    _classCallCheck(this, Core);

    var router = options.router,
        store = options.store,
        _options$locales = options.locales,
        locales = _options$locales === void 0 ? {} : _options$locales,
        login = options.login,
        layout = options.layout,
        moduleConfig = options.moduleConfig,
        _options$theme = options.theme,
        theme = _options$theme === void 0 ? 'ios' : _options$theme,
        _options$HomeConfig = options.HomeConfig,
        HomeConfig = _options$HomeConfig === void 0 ? {
      needBottomBar: true
    } : _options$HomeConfig,
        nextHomeConfig = options.nextHomeConfig,
        _options$saas = options.saas,
        saas = _options$saas === void 0 ? false : _options$saas;
    this.router = router;
    this.store = store;
    this.locales = locales;
    this.login = login ? login : require("./components/Login");
    this.layout = layout ? layout : require("./components/Layout");
    this.moduleConfig = moduleConfig;
    this.theme = theme;
    isSaas = saas;
    HomeConfigs = HomeConfig || {};
    NextConfigs = nextHomeConfig || {};

    _toolkit.session.setSaas(saas);
  }

  _createClass(Core, [{
    key: "getReducers",
    value: function getReducers() {
      var moduleReducers = modules.reduce(function (p, c) {
        return _objectSpread({}, p, {}, c.store);
      }, {});
      var loginReducers = this.login.store;
      var layoutReducers = this.layout.store;
      return _objectSpread({}, moduleReducers, {}, loginReducers, {}, layoutReducers, {}, this.store);
    } // 先合并数组 然后再追加数组。经过测试数组长度为30时,执行的时长为1ms。因此次算法的性能基本可以满足。

  }, {
    key: "concatNotSort",
    value: function concatNotSort(arr1, arr2) {
      var newArr = [].concat(_toConsumableArray(arr1), _toConsumableArray(arr2));
      var pathArr = [].concat(_toConsumableArray(arr1.map(function (item) {
        return item.path;
      })), _toConsumableArray(arr2.map(function (item) {
        return item.path;
      })));
      var updateArr = [];
      pathArr.forEach(function (item, index) {
        if (updateArr.map(function (item) {
          return item.path;
        }).indexOf(item) === -1) {
          updateArr.push(newArr[index]);
        } else {
          var i = updateArr.map(function (item) {
            return item.path;
          }).indexOf(item);
          delete updateArr[i];
          updateArr.push(newArr[index]);
        }
      });
      return updateArr;
    }
  }, {
    key: "initPlugins",
    value: function initPlugins() {
      var _registerPlugins = (0, _plugin.registerPlugins)(),
          routers = _registerPlugins.routers,
          stores = _registerPlugins.stores,
          locales = _registerPlugins.locales;

      this.pluginRouter = routers;
      this.pluginStore = stores;
      this.pluginLocales = locales;
    }
  }, {
    key: "getRouters",
    value: function getRouters() {
      var moduleRoutes = (0, _lodash.flattenDeep)(modules.map(function (m) {
        return m.router;
      }));

      function checkAuth(to, from, resolve, reject) {
        var router = this;

        if (_toolkit.session.isLoggedIn()) {
          resolve();
        } else {
          reject();
          router.navigate('/login');
        }
      }

      var routers = this.concatNotSort(moduleRoutes, this.router).map(function (r) {
        if ((0, _lodash.isArray)(r.beforeEnter)) {
          r.beforeEnter = [].concat(_toConsumableArray(r.beforeEnter), [checkAuth]);
        } else if (r.beforeEnter) {
          r.beforeEnter = [r.beforeEnter, checkAuth];
        } else {
          r.beforeEnter = checkAuth;
        }

        return r;
      });
      return routers;
    }
  }, {
    key: "setLocales",
    value: function setLocales() {
      modules.forEach(function (m) {
        _toolkit.i18n.setLocales(m.locale);
      });

      _toolkit.i18n.setLocales(this.locales);
    } // func mount

  }, {
    key: "mount",
    value: function mount(selector) {
      var _this = this;

      var initialState = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      this.initPlugins();

      _framework7LiteEsmBundle.default.use(_fUi.default);

      this.setLocales();

      var locale = _toolkit.i18n.getLocalSettings();

      var store = (0, _store.default)(this.getReducers(), initialState);

      _toolkit.i18n.load(locale).then(function () {
        _reactDom.default.render(_react.default.createElement(_reactRedux.Provider, {
          store: store
        }, _react.default.createElement(_Context.default.Provider, {
          value: _this.moduleConfig
        }, _react.default.createElement(_Root.default, {
          routes: _this.getRouters(),
          theme: _this.theme
        }))), document.querySelector(selector));
      });
    }
  }]);

  return Core;
}();

exports.default = Core;

_defineProperty(Core, "use", function (module) {
  modules.push(module);
});

_defineProperty(Core, "getSaas", function () {
  return isSaas;
});

_defineProperty(Core, "plugin", function (p) {
  return (0, _plugin.addPlugin)(p);
});

_defineProperty(Core, "getHomeConfig", function () {
  return HomeConfigs;
});

_defineProperty(Core, "getNextHomeConfig", function () {
  return NextConfigs;
});
